src/main/scala/vsys/blockchain/state/opcdiffs/CDBVOpcDiff.scala

```diff
@@ -9,33 +9,30 @@ import vsys.blockchain.contract.Contract.checkStateVar
 
 import scala.util.{Left, Right, Try}
 
-object CDBVOpcDiff {
+object CDBVOpcDiff extends OpcDiffer {
 
   def set(context: ExecutionContext)(stateVar: Array[Byte],
                                      value: DataEntry): Either[ValidationError, OpcDiff] = {
-    if (!checkStateVar(stateVar, value.dataType)) {
-      Left(ContractInvalidStateVariable)
-    } else {
-      if (value.dataType == DataType.Address) {
-        val a = Address.fromBytes(value.data).toOption.get
-        Right(OpcDiff(relatedAddress = Map(a -> true),
-          contractDB = Map(ByteStr(context.contractId.bytes.arr ++ Array(stateVar(0))) -> value.bytes)))
-      } else {
-        Right(OpcDiff(contractDB = Map(ByteStr(context.contractId.bytes.arr
-          ++ Array(stateVar(0))) -> value.bytes)))
+    for {
+      _ <- Either.cond(checkStateVar(stateVar, value.dataType), (), ContractInvalidStateVariable)
+      addrOpt <- if (value.dataType == DataType.Address) Address.fromBytes(value.data).map(Some(_))
+              else Right(None)
+      contractDB = Map(ByteStr(context.contractId.bytes.arr ++ Array(stateVar(0))) -> value.bytes)
+      diff = addrOpt match {
+        case Some(addr) => OpcDiff(relatedAddress = Map(addr -> true), contractDB = contractDB)
+        case None => OpcDiff(contractDB = contractDB)
       }
-    }
+    } yield diff
```

Comment:
the original logic is

|type | address valid| result |
|:-:|:-:|:-:|
| address |valid | diff with relatedAddress |
|address|invalid| error|
|others|/|diff only|

now

|type | address valid| result |
|:-:|:-:|:-:|
| address |valid | diff with relatedAddress |
|address|invalid| diff only|
|others|/|diff only|

seems different in logic. if require type is address, but get an invalid address, it should return validation error? the one why it can use get directly is that, the address type validation is checked when built the data entry.

Reply:
now invalid address is still error. `Address.fromBytes` returns `Either`, `.map` only map the Right

Reply:
i confused the address check within if. and also here validation error is the left~